src/Storages/StorageS3.cpp

```diff
@@ -327,6 +345,167 @@ class StorageS3Sink : public SinkToStorage
+
+        const auto * key_column = checkAndGetColumn<ColumnString>(block_with_partition_by_expr.getByName(partition_by_column_name).column.get());
+
+        std::unordered_map<String, size_t> sub_chunks_indices;
+        IColumn::Selector selector;
+        for (size_t row = 0; row < chunk.getNumRows(); ++row)
+        {
+            auto value = key_column->getDataAt(row);
+            validatePartitionKey(value);
+            auto [it, inserted] = sub_chunks_indices.emplace(value, sub_chunks_indices.size());
+            selector.push_back(it->second);
+        }
+
+        Chunks sub_chunks;
+        sub_chunks.reserve(sub_chunks_indices.size());
+        for (size_t column_index = 0; column_index < columns.size(); ++column_index)
+        {
+            MutableColumns column_sub_chunks = columns[column_index]->scatter(sub_chunks_indices.size(), selector);
+            if (column_index == 0) /// Set sizes for sub-chunks.
+            {
+                for (size_t sub_chunk_index = 0; sub_chunk_index < column_sub_chunks.size(); ++sub_chunk_index)
+                {
+                    sub_chunks.emplace_back(Columns(), column_sub_chunks[sub_chunk_index]->size());
+                }
+            }
+            for (size_t sub_chunk_index = 0; sub_chunk_index < column_sub_chunks.size(); ++sub_chunk_index)
+            {
+                sub_chunks[sub_chunk_index].addColumn(std::move(column_sub_chunks[sub_chunk_index]));
+            }
+        }
+
+        for (const auto & [partition_id, sub_chunk_index] : sub_chunks_indices)
+        {
+            getSinkForPartition(partition_id)->consume(std::move(sub_chunks[sub_chunk_index]));
+        }
+    }
+
+    void onFinish() override
+    {
+        for (auto & [partition_id, sink] : sinks)
+        {
+            sink->onFinish();
+        }
+    }
+
+private:
+    using SinkPtr = std::shared_ptr<StorageS3Sink>;
+
+    const String format;
+    const Block sample_block;
+    ContextPtr context;
+    const CompressionMethod compression_method;
+    std::shared_ptr<Aws::S3::S3Client> client;
+    const String bucket;
+    const String key;
+    size_t min_upload_part_size;
+    size_t max_single_part_upload_size;
+
+    ExpressionActionsPtr partition_by_expr;
+    String partition_by_column_name;
+
+    std::unordered_map<String, SinkPtr> sinks;
+
+    static String replaceWildcards(const String & haystack, const String & partition_id)
+    {
+        return boost::replace_all_copy(haystack, PARTITION_ID_WILDCARD, partition_id);
+    }
+
+    SinkPtr getSinkForPartition(const String & partition_id)
+    {
+        if (sinks.count(partition_id) == 0)
+        {
+            sinks.emplace(partition_id, std::make_shared<StorageS3Sink>(
+                format,
+                sample_block,
+                context,
+                compression_method,
+                client,
+                replaceWildcards(bucket, partition_id),
+                replaceWildcards(key, partition_id),
```

Comment:
Should we check that result bucket name and key are not empty? And also maximal length?

Reply:
@vdimir Bucket name is being checked in `S3::URI`, and empty key is allowed in AWS.

Reply:
Do you mean that it's checked on server-side?

```
INSERT INTO FUNCTION s3('http://localhost:9001/foo/{_partition_id}', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val VALUES

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. (S3_ERROR)
```


```
INSERT INTO FUNCTION s3('http://localhost:9001/{_partition_id}/key.csv', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val VALUES

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. (S3_ERROR)
```